include/rocksdb/write_buffer_manager.h

```diff
@@ -56,121 +64,132 @@ class WriteBufferManager final {
-  // pass allow_stall = true during WriteBufferManager instance creation.
+  // Returns true if stall conditions are met.
+  // Stall conditions: stall is allowed AND memory limit is set (i.e, buffer
+  // size > 0) AND total memory usage accounted by this WriteBufferManager
+  // exceeds the memory limit.
+  //
+  // WARNING: Should only be called by RocksDB internally .
   //
-  // Should only be called by RocksDB internally .
+  // WARNING: If running without syncronization with any functions that could
+  // change the stall conditions above, this function might not return the
+  // latest result changed by these functions but an old result.
   bool ShouldStall() const {
-    if (!allow_stall_ || !enabled()) {
+    if (allow_stall_.load(std::memory_order_relaxed) && enabled() &&
+        IsStallThresholdExceeded()) {
+      return true;
+    } else {
       return false;
     }
-
-    return IsStallActive() || IsStallThresholdExceeded();
-  }
-
-  // Returns true if stall is active.
-  bool IsStallActive() const {
-    return stall_active_.load(std::memory_order_relaxed);
-  }
-
-  // Returns true if stalling condition is met.
-  bool IsStallThresholdExceeded() const {
-    return memory_usage() >= buffer_size_;
   }
 
+  // WARNING: Should only be called by RocksDB internally.
   void ReserveMem(size_t mem);
 
   // We are in the process of freeing `mem` bytes, so it is not considered
   // when checking the soft limit.
+  //
+  // WARNING: Should only be called by RocksDB internally.
   void ScheduleFreeMem(size_t mem);
 
+  // WARNING: Should only be called by RocksDB internally.
   void FreeMem(size_t mem);
 
-  // Add the DB instance to the queue and block the DB.
-  // Should only be called by RocksDB internally.
-  void BeginWriteStall(StallInterface* wbm_stall);
+  // Return true if stall conditions have been met and WriteBufferManager has
+  // done its work to begin write stall. Return false otherwise.
+  //
+  // WARNING: Should only be called by RocksDB internally.
+  bool MaybeBeginWriteStall(StallInterface* wbm_stall);
 
+  // WARNING: Should only be called by RocksDB internally.
+  void RemoveDBFromQueue(StallInterface* wbm_stall);
+
+ private:
   // If stall conditions have resolved, remove DB instances from queue and
   // signal them to continue.
+  //
+  // Called when stall conditions (see `ShouldStall()` API) might have been
+  // changed
+  //
+  // REQUIRED: wbm_mutex_ held
   void MaybeEndWriteStall();
 
-  void RemoveDBFromQueue(StallInterface* wbm_stall);
+  // REQUIRED: wbm_mutex_ held
+  bool IsStallThresholdExceeded() const {
+    return memory_usage() >= buffer_size_;
+  }
+
+  // WARNING:  Should only be called from write thread
+  // REQUIRED: wbm_mutex_ held
+  void ReserveMemWithCache(size_t mem);
+
+  // REQUIRED: wbm_mutex_ held
+  void FreeMemWithCache(size_t mem);
+
+  // Mutex used to protect WriteBufferManager's data variables.
+  mutable std::mutex wbm_mutex_;
```

Comment:
Cleanup 6: use 1 instead of multiple mutex to guard separate groups of WBM data for simplicity.